zerver/lib/outgoing_webhook.py

```diff
@@ -0,0 +1,102 @@
+from __future__ import absolute_import
+from typing import Any, Iterable, Dict, Tuple, Callable, Text, Mapping
+
+import requests
+import json
+import sys
+import inspect
+import logging
+from six.moves import urllib
+from functools import reduce
+
+from django.utils.translation import ugettext as _
+
+from zerver.models import Realm, get_realm_by_email_domain, get_user_profile_by_id
+from zerver.lib.actions import internal_send_message
+from zerver.lib.queue import queue_json_publish
+from zerver.lib.validator import check_dict, check_string
+from zerver.decorator import JsonableError
+
+class ServiceMessageActions(object):
+
+    @staticmethod
+    def send_response_message(realm, bot_email, trigger_message, response_message_content):
+        # type: (Realm, Text, Dict[str, Any], Text) -> None
+        recipient_type_name = trigger_message['type']
+        if recipient_type_name == 'stream':
+            recipients = trigger_message['display_recipient']
+            internal_send_message(realm, bot_email, recipient_type_name, recipients,
+                                  trigger_message['subject'], response_message_content)
+        else:
+            # Private message; only send if the bot is there in the recipients
+            trigger_message_recipients = [recipient['email'] for recipient in trigger_message['display_recipient']]
+            if bot_email in trigger_message_recipients:
+                recipients = ','.join(trigger_message_recipients)
+                internal_send_message(realm, bot_email, recipient_type_name, recipients,
+                                      trigger_message['subject'], response_message_content)
+
+    @classmethod
+    def succeed_with_message(cls, event, success_message):
+        # type: (Dict[str, Any], Text) -> None
+        bot_user = get_user_profile_by_id(event['user_profile_id'])
+        trigger_message = event['message']
+        realm = get_realm_by_email_domain(event['message']['sender_email'])
+        cls.send_response_message(realm, bot_user.email, trigger_message, success_message)
```

Comment:
I think we should pass the `bot_user` into these methods, rather than having them look it up each time.  It'll save a bunch of unnecessary queries and complexity tracing the code path.